dnshostprovider.go

```diff
@@ -2,25 +2,55 @@ package zk
 
 import (
 	"fmt"
+	"log/slog"
+	"math/rand"
 	"net"
 	"sync"
 )
 
-// DNSHostProvider is the default HostProvider. It currently matches
-// the Java StaticHostProvider, resolving hosts from DNS once during
-// the call to Init.  It could be easily extended to re-query DNS
-// periodically or if there is trouble connecting.
+type hostPort struct {
+	host, port string
+}
+
+func (hp *hostPort) String() string {
+	return hp.host + ":" + hp.port
+}
+
+// DNSHostProvider is the default HostProvider. It currently supports two modes: dynamic or static
+// resolution of the addresses backing the ZK hosts.
+//
+// When in dynamic resolution mode, it will iterate through the ZK hosts on every call to Next, and
+// return a random address selected from the resolved addresses of the ZK host (if the host is
+// already an IP, it will return that directly). It is important to manually resolve and shuffle the
+// addresses because the DNS record that backs a host may rarely (or never) change, so repeated calls
+// to connect to this host may always connect to the same IP. This mode is the default mode, and
+// matches the Java client's implementation. Note that if the host cannot be resolved, Next will
+// return it directly, instead of an error. This will cause Dial to fail and the loop will move on to
+// a new host.
+// https://github.com/linkedin/zookeeper/blob/629518b5ea2b26d88a9ec53d5a422afe9b12e452/zookeeper-server/src/main/java/org/apache/zookeeper/client/StaticHostProvider.java#L368
+//
+// When in static resolution mode, it resolves the hosts once during Init, and iterates through the
+// resolved addresses for every call to Next. Note that if the addresses that back the ZK hosts
+// change, those changes will not be reflected. This mode is preserved to support previous behavior,
+// and is not the default.
 type DNSHostProvider struct {
+	static bool
+
 	mu         sync.Mutex // Protects everything, so we can add asynchronous updates later.
-	servers    []string
+	servers    []hostPort
 	curr       int
 	last       int
 	lookupHost func(string) ([]string, error) // Override of net.LookupHost, for testing.
 }
 
-// Init is called first, with the servers specified in the connection
-// string. It uses DNS to look up addresses for each server, then
-// shuffles them all together.
+func NewDynamicDNSHostProvider() HostProvider {
+	return &DNSHostProvider{static: false}
+}
+
+func NewStaticDNSHostProvider() HostProvider {
+	return &DNSHostProvider{static: true}
+}
```

Comment:
Can a new implementation of the `DNSHostProvider` interface be created rather than modifying existing code? If possible, it should be preferred because of safety and readability.

Reply:
I did it that way since they were fundamentally not that different but sure, I'll copy it and mark the old one as deprecated. The semantics of the `HostProvider` interface have changed though (`Next` returns a boolean on call earlier), so there will still be some changes.

Reply:
Done in #6